File: TypeScript/src/roman_to_integer.ts
function romanToInt(s: string): number {
  const romanMap = new Map<string, number>([
    ["I", 1],
    ["V", 5],
    ["X", 10],
    ["L", 50],
    ["C", 100],
    ["D", 500],
    ["M", 1000],
  ]);

  let result = 0;

  for (const char of s.split("").reverse()) {
    const number = romanMap.get(char);

    if (number === undefined) {
      return -1;
    }

    if (number * 4 < result) {
      result -= number;
    } else {
      result += number;
    }
  }

  return result;
}

export { romanToInt };

File: Ruby/lib/roman_to_integer.rb
# @param {String} s
# @return {Integer}
def roman_to_int(s)
  roman_map = {
    I: 1,
    V: 5,
    X: 10,
    L: 50,
    C: 100,
    D: 500,
    M: 1000,
  }

  result = 0

  s.reverse.each_char do |c|
    number = roman_map[c.to_sym]
    return -1 if number == nil

    if number * 4 < result
      result -= number
    else
      result += number
    end
  end

  return result
end
